components/omega/doc/design/Decomp.md

```diff
@@ -0,0 +1,260 @@
+decomposition.
+
+
+## 2 Requirements
+
+### 2.1 Requirement: Mesh partition
+
+The primary requirement is to be able to partition a mesh across a
+distributed memory machine. Many standard tools exist to perform
+this partitioning (eg metis, parmetis, zoltan) and their use is
+encouraged. This normally requires an input file with the primary
+mesh indices and adjacency (see Requirement 2.4).
+
+### 2.2 Requirement: Metis option
+
+For backward compatibility and to facilitate bit-for-bit comparisons
+with MPAS, at least one partition option must use the Metis library.
+
+### 2.3 Required: Online partitioning
+
+We require the partitioning to occur online (ie as part of model 
+initialization rather than a separate pre-processing step).
+This is meant to avoid the past difficulties in maintaining a large
+number of partition input files for each MPI configuration. Experience
+has shown that even high-resolution partitioning is very inexpensive
+and can be performed as part of the model run and does not need this
+off-line step.
+
+### 2.4 Required: Input mesh information
+
+As input into the partitioning, we require a file with information
+on cell adjacency. There are at least two potential design approaches
+for this.  The first option is to use the global `graph.info` file
+which is already in Metis input format with a line for each cell 
+listing the global indices of that cell's neighbors. A second option 
+is to use the input mesh netCDF file which has the same information
+in the global cellsOnCell array. This second option would enable us to
+eliminate all graph.info files after the mesh file has been created. 
+
+### 2.5 Required: All mesh locations
+
+While the primary partition is based on the cells and cell
+adjacency, the decomposition must also partition the edge and vertex
+index spaces.
+
+### 2.6 Required: Local to global index mapping
+
+The decomposition must provide the ability to return the global
+index given the processor's local index. It must also be able to
+provide the location (MPI rank and local index) for all the
+neighbor or halo points to facilitate later setup of halos and
+other infrastructure. Some of this information may only be needed
+for initialization (see Requirement 2.8 on release of memory).
+
+### 2.7 Required: Local sorting
+
+At a minimum, the local addresses must be sorted with owned
+indices first and each level of halo following. This is needed
+for both optimal communication and for selectively computing
+only what is needed. Additional sorting within those blocks
+may be desireable for optimizing memory locality.
+
+### 2.8 Required: Release memory
+
+Once the partition information has been used to set up halos,
+mesh variables, I/O and other quantities that depend on the
+partition, all or parts of the partitioning data may be
+released to free memory. Note that some of the information
+will be transferred to mesh variables as part of later
+mesh initialization.
```

Comment:
Please add a requirement that the local partition includes a halo, where the halo width may be specified in the input deck. Do we want to specify that the local indexing has the same ordering as MPAS now, with the halo rings at the end in consecutive order?

Reply:
Added the halo requirement